src/rqt_msg/messages_widget.py

```diff
@@ -211,34 +215,48 @@ def _rightclick_menu(self, event):
             self._logger.debug('_rightclick_menu selected={}'.format(selected))
             selected_type = selected[1].data()
 
-            if selected_type.find('[') >= 0:
-                selected_type = selected_type[:selected_type.find('[')]
+            # We strip any array information for loading the python classes for messages, services and actions
+            selected_type_bare = selected_type
+            if selected_type_bare.find('[') >= 0:
+                selected_type_bare = selected_type_bare[:selected_type_bare.find('[')]
+
+            # We use the number of '/' to determine of the selected type is a msg, action, service, or primitive type
+            selected_type_bare_tokens_len = len(selected_type_bare.split('/'))
 
-            # TODO(mlautman):
-            #   implement get_msg_text like functionality like what is available in ROS1
             browsetext = None
 
             if (self._mode == message_helpers.MSG_MODE):
```

Comment:
This if block is getting a bit unwieldy and difficult to track the logic. Could it be abstracted into separate methods?

Or maybe when the tokens_len is 1 or 2 that gets pulled out of this if block since that logic is all similar.

Reply:
I refactored this and fixed another bug. I also added a TODO to stop using the number of '/' in the name to determine if the item is a message, service or action once there is a better way available